src/hpr_packet_reporter.erl

```diff
@@ -0,0 +1,432 @@
+            file:write(S, PacketSize),
+            file:write(S, Packet)
+        end,
+        Data
+    ),
+    file:close(S),
+
+    NumDeleted = delete_packets_by_timestamp(WriteTimestamp),
+    FileSize = filelib:file_size(FilePath),
+
+    lager:info(
+        [
+            {packets_processed, length(Data)},
+            {packets_deleted, NumDeleted},
+            {file_size, FileSize},
+            {max_file_size, MaxFileSize}
+        ],
+        "packet reporter processing"
+    ),
+
+    {ok, FileSize}.
+
+-spec setup_aws(packet_reporter_opts()) -> aws_client:aws_client().
+setup_aws(#{
+    access_key_id := AccessKey,
+    secret_access_key := Secret,
+    region := Region,
+    localstack_port := LocalstackPort,
+    localstack_host := LocalstackHost
+}) ->
+    case Region of
+        <<"local">> ->
+            aws_client:make_local_client(
+                AccessKey,
+                Secret,
+                LocalstackPort,
+                LocalstackHost
+            );
+        _ ->
+            aws_client:make_client(AccessKey, Secret, Region)
+    end.
+
+-spec start_report_interval(IntervalDuration :: interval_duration_ms()) -> {ok, timer:tref()}.
+
+start_report_interval(IntervalDuration) ->
+    timer:apply_interval(IntervalDuration, ?MODULE, write_packets, []).
+
+-spec handle_restart_interval(
+    IntervalRef :: timer:tref() | undefined, IntervalDuration :: interval_duration_ms()
+) -> {ok, timer:tref()}.
+handle_restart_interval(undefined, IntervalDuration) ->
+    start_report_interval(IntervalDuration);
+handle_restart_interval(IntervalRef, IntervalDuration) ->
+    timer:cancel(IntervalRef),
+    start_report_interval(IntervalDuration).
+
+-spec handle_stop_interval(IntervalRef :: timer:tref()) -> {ok, cancel} | {error, term()}.
+handle_stop_interval(undefined) -> {ok, no_interval};
+handle_stop_interval(IntervalRef) -> timer:cancel(IntervalRef).
+
+-spec encode_packet(Packet :: hpr_packet_up:packet(), PacketRoute :: hpr_route:route()) -> binary().
+encode_packet(
+    #packet_router_packet_up_v1_pb{
+        payload = Payload,
+        timestamp = GatewayTimestamp,
+        rssi = RSSI,
+        frequency = FrequencyMhz,
+        datarate = Datarate,
+        snr = SNR,
+        region = Region,
+        gateway = Gateway
+    },
+    #config_route_v1_pb{
+        oui = OUI,
+        net_id = NetID
+    }
+) ->
+    packet_router_pb:encode_msg(
+        #packet_router_packet_report_v1_pb{
+            gateway_timestamp_ms = GatewayTimestamp,
+            oui = OUI,
+            net_id = NetID,
+            rssi = RSSI,
+            frequency = FrequencyMhz,
+            datarate = Datarate,
+            snr = SNR,
+            region = Region,
+            gateway = Gateway,
+            payload_hash = crypto:hash(sha256, Payload)
+        },
+        packet_router_packet_report_v1_pb
+    ).
+
+-spec encode_packet_size(EncodedPacket :: binary()) -> PacketSize :: binary().
+encode_packet_size(EncodedPacket) ->
```

Comment:
Is this the same as `<<PacketSize:32/big-integer-unsigned>>`?

Reply:
Very cool, changing.